### PR TITLE
fixed the specs apart from scars - but have raised some questions

### PR DESCRIPTION
**Pull request by kingsleyh**
I've fixed all the specs apart the scars specs - and some things mentioned in my notes below. here are my notes on some things I found:

1. spec/units/blockchain/transaction.cr:51
   A new check has been added for prev_hash (transaction:53) which causes an invalid signing error for some reason lower down when the prev_hash is the expected one.

2. spec/units/blockchain/transaction.cr:199
   This test checked there was 1 sender. The code has removed this check but I think it should be put back in as we use senders.first lower down in transaction.cr to get the 
   public_key for signing

3. spec/units/node/rpc_controller.cr:41
   There is no check for invalid fee for an unsigned transaction now.
   if no invalid fee error is raised when creating an unsigned transaction - at what
   point will this check occur? If a user sends a transaction with a fee which is lower than than the specified fee for
   the transaction - how will they get to know that their transaction was rejected for this reason? Having this check here was handy to avoid
   simple mistakes with the fee right at the point of sending the transaction.

4. spec/units/blockchain/scars.cr 
   The specs look completely different to the scars.cr implementation - I guess they need to be re-written to match the current implementation?

5. dApps specs
   We should write tests that cover indices

### CODE REVIEW

**Comment by kingsleyh** [comment]
Do you want me to re-write the specs for scars? Or you going to do it?

**Comment by tbrand** [comment]
> (3) There is no check for invalid fee for an unsigned transaction now.

Yeah, I do it in cli. And I have a plan to create a rejected transactions pool with the reason as dApps.

> (4) The specs look completely different to the scars.cr implementation

Oops, I forgot to remove it. Sorry.
Now I've removed it, please sync again.

> (5) dApps specs We should write tests that cover indices

Yeah. Or we can do it later.

> Do you want me to re-write the specs for scars? Or you going to do it?

Leave it to you :pray:

File: spec/units/blockchain/consensus.cr [hunk 1]
```diff
@@ -28,11 +29,11 @@ describe Consensus do
 
   describe "difficulty" do
     it "should return the #difficulty_at for the blockchain" do
-      difficulty_at(0_i64).should eq(4)
+      difficulty_at(0_i64).should eq(3)
```
Comment (tbrand) [inline spec/units/blockchain/consensus.cr]: Oops, that's my bad! 😢 
I set a easy difficulty for the debugging. It should be same as before :pray:
I'll revert here!

File: spec/units/blockchain/block.cr [hunk 2]
```diff
@@ -4,7 +4,7 @@ require "./../utils"
 include Sushi::Core
 include Hashes
 include Units::Utils
-include Sushi::Common::Num
+# include Sushi::Common::Num
```
Comment (tbrand) [inline spec/units/blockchain/block.cr]: Num module is remove completely.

Comment (kingsleyh) [reply spec/units/blockchain/block.cr]: have removed it now

File: spec/units/blockchain/consensus.cr (hunk 1 above)
Comment (kingsleyh) [reply spec/units/blockchain/consensus.cr]: cool

**Comment by kingsleyh** [comment]
It will take another couple of days to finish the scars specs - but I've made great progress by writing the chain_generator

File: spec/units/utils/chain_generator.cr [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    property miner_wallet : Wallet
+    property transaction_factory : TransactionFactory
+
+    def initialize
+      @node_wallet = Wallet.from_json(Wallet.create(true).to_json)
+      @miner_wallet = Wallet.from_json(Wallet.create(true).to_json)
+      @blockchain = Blockchain.new(node_wallet)
+      @miner = {address: miner_wallet.address, socket: MockWebSocket.new, nonces: [] of UInt64}
+      @transaction_factory = TransactionFactory.new(@node_wallet)
+      ENV["UT"] = "unit test"
```
Comment (tbrand) [inline spec/units/utils/chain_generator.cr]: Where are you using it?

File: src/core/blockchain/dapps/scars.cr [hunk 4]
```diff
@@ -86,23 +86,24 @@ module ::Sushi::Core
 
       recipient = transaction.recipients[0]
 
-      raise "address mistach for scars_sell: #{address} vs #{recipient[:address]}" if address != recipient[:address]
```
Comment (tbrand) [inline src/core/blockchain/dapps/scars.cr]: Thanks for fixing the message!:+1::+1::+1:

**Comment by kingsleyh** [comment]
All done - you can review and merge into your scars branch when you have time. Then hopefully merge into master :)

**Comment by tbrand** [comment]
Awesome works! Thanks a lot!!!

**Pull request merged**